lib/agent.js

```diff
@@ -0,0 +1,215 @@
+'use strict'
+
+const {
+  ClientClosedError,
+  InvalidArgumentError,
+  ClientDestroyedError
+} = require('./core/errors')
+const Dispatcher = require('./dispatcher')
+const Pool = require('./pool')
+const Client = require('./client')
+const util = require('./core/util')
+const assert = require('assert')
+const RedirectHandler = require('./handler/redirect')
+
+const kDestroyed = Symbol('destroyed')
+const kClosed = Symbol('closed')
+const kOnConnect = Symbol('onConnect')
+const kOnDisconnect = Symbol('onDisconnect')
+const kClients = Symbol('clients')
+const kMaxRedirections = Symbol('maxRedirections')
+const kOnDrain = Symbol('onDrain')
+const kFactory = Symbol('factory')
+const kFinalizer = Symbol('finalizer')
+const kOptions = Symbol('options')
+
+function defaultFactory (origin, opts) {
+  return opts && opts.connections === 1
+    ? new Client(origin, opts)
+    : new Pool(origin, opts)
+}
+
+const WeakRef = global.WeakRef || class DummyWeakRef {
+  constructor (value) {
+    this.value = value
+  }
+
+  deref () {
+    return this.value
+  }
+}
+const FinalizationRegistry = global.FinalizationRegistry || class DummyFinalizer {
+  register () {}
+}
```

Comment:
Can you please add a comment explaining why these are ok and do not leak memory?

Reply:
It does leak a little memory.  See, https://github.com/nodejs/undici/pull/620#discussion_r603319160.

If that's unacceptable I will have to try and find more complicate solutions.

Reply:
Basically it will keep a few object refs for every origin it sees. Which shouldn't be that many in practice.